### DISCUSSION TITLE
Custom error messages in case of failure

### DISCUSSION THREAD

**Discussion by MustafaI** [j-mie6/Parsley#69] [Q&A] [answered]
I'd like to use parsley to create custom error messages in the case where runParser returns Failure(..). From what I can see it's currently not possible to retrieve error line, column, expected and unexpected strings and so on without modifying parsley.internal.instructions.Context. Would I be correct in saying this? If so, are there any plans to implement this into Parsley in future?

Thanks.

**Comment by j-mie6** [comment, marked as answer]
Yes, Parsley 3 will return the `ParseError` datatype instead of a string, along with a pretty printing typeclass

**Reply by j-mie6** [reply]
But since Parsley 3 will include a few breaking changes, I don't plan to release it for a few weeks at least.

Perhaps there is another way of achieving what you want?

**Reply by MustafaI** [reply]
Thanks for the quick response!

If it'll take a while to release then that's fine, I should be able to modify parsley to give me what I need in meantime.

**Reply by j-mie6** [reply]
Ok! You should find it easy enough: you need to modify the `Context` so it doesn't call `.pretty`, then make the `ParseError` stuff in parsley.internal.instructions.Errors not `private [instructions]`

**Reply by MustafaI** [reply]
@j-mie6 A quick follow up question on this, in the case of runParser returning Success(..), would the private fields "line" and "col" in parslet.internal.instructions.Context refer to the start of the text that the parser matched? So if we were parsing for hello in "abcdefhello" would "col" be 6/7 (depending on if col starts at zero)?

**Reply by j-mie6** [reply]
it would be the _current_ position at the point of success. Your parser _should_ ideally have `eof` at the end of it to ensure that you read the entire file, so it'll normally just literally be the end of the file's line (and column + 1). On the other hand, the `pos` combinator can produce an `(Int, Int)` at any point whilst the parser is running. What do you hope to achieve with the line and column?

**Reply by MustafaI** [reply]
The `pos` combinator is exactly what I needed. Thank you.
